sematic/db/queries.py

```diff
@@ -139,6 +144,91 @@ def save_run(run: Run) -> Run:
     return run
 
 
+def save_external_resource_record(record: ExternalResourceRecord):
+    """Save an ExternalResourceRecord to the DB"""
+    existing_record = get_external_resource_record(record.id)
+
+    if existing_record is None:
+        if record.resource_state != ResourceState.CREATED:
+            raise ValueError(
+                f"Cannot create an external resource in a state other than 'CREATED'. "
+                f"{record.id} was in state '{record.resource_state}'."
+            )
+
+        # this ensures that all the fields are consistent
+        record = ExternalResourceRecord.from_resource(record.resource)
+    else:
+        # this ensures that the update properly updates history
+        # and keeps data consistent
+        existing_record.resource = record.resource
+        record = existing_record
+
+    with db().get_session() as session:
+        session.merge(record)
+        session.commit()
+
+        return record
+
+
+def get_external_resource_record(resource_id: str) -> Optional[ExternalResourceRecord]:
+    """Get an ExternalResourceRecord from the DB"""
+    with db().get_session() as session:
+        return (
+            session.query(ExternalResourceRecord)
+            .filter(ExternalResourceRecord.id == resource_id)
+            .one_or_none()
+        )
+
+
+def save_run_external_resource_links(resource_ids: List[str], run_id: str):
+    """Save the relationship between external resources and a run to the DB."""
+    with db().get_session() as session:
+        for resource_id in resource_ids:
+            session.merge(RunExternalResource(resource_id=resource_id, run_id=run_id))
+        session.commit()
+
+
+def get_resource_ids_by_root_id(root_run_id: str) -> List[str]:
+    """Get a list of ids of external resources associated with a particular root run"""
+    with db().get_session() as session:
+        results = (
+            session.query(ExternalResourceRecord.id)
+            .filter(ExternalResourceRecord.id == RunExternalResource.resource_id)
+            .filter(Run.id == RunExternalResource.run_id)
+            .filter(Run.root_id == root_run_id)
+            .all()
```

Comment:
I believe this might solve the 2-query problem:
```suggestion
            session.query(ExternalResourceRecord)
            .join(RunExternalResource, ExternalResourceRecord.id == RunExternalResource.resource_id)
            .join(Run, Run.id == RunExternalResource.run_id)
            .filter(Run.root_id == root_run_id)
            .distinct()
            .all()
```

Reply:
will do!